### PR TITLE
Implement Unity Concord

### PR DESCRIPTION
**Pull request by ghost**
**_I affirm:_**
- [x] that I agree to Topaz Next's [Limited Contributor License Agreement](https://github.com/topaz-next/topaz/blob/release/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have read the [Contributing Guide](https://github.com/topaz-next/topaz/blob/release/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/topaz-next/topaz/blob/release/CODE_OF_CONDUCT.md)
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

**Core**
- [x] Unity Weekly and Partial Rankings
- [x] Allow for joining and changing Unity faction
- [x] Enable or Disable Unity Leader Trust based on personal ranking
- [x] Add support for Synthesis objectives based on crafting type
- [x] Add support for Weapon Skill objectives to receive skill type data
- [x] Add support for Heal Unity Ally objectives
- [x] Add `roeutils::event()` binding (needed for HELM, Magic Bursts, and Chocobo Digging, could be made specific)
- [x] Implements Unity Chat and RoE Shared Objectives

**Lua:**
- [x] Support Joining and Changing faction with Unity NPC
- [x] Unity Teleports (few missing)
- [x] Add support for Harvesting objectives
- [x] Add support for Chocobo Digging objectives
- [x] Add support for Magic Burst Objectives
- [x] Purchase non-fight related items

**RoE Objectives:**
- [x] Add supported Harvesting Objectives
- [x] Add supported Chocobo Digging objectives
- [x] Add supported Unity Shared Objectives (Missing HP, Fishing, Monstrosity, and Vanquish w/ Leader)
- [x] Add supported Unity Leader Objectives (Missing Cure Status Ailments)

What this does **_NOT_** do:
* Unity Concord Fights
* Upgrade gear from Unity Concord Fights
* Implement Unity Trust Gambits

**Known Issues:**
* Unity chat member limit is uncapped.  This is intentional due to the possibility that a total value will be inaccurate when running multiple map servers.

**Test Scenarios NOT covered yet**
* Accolade granted from Objective without being a member

### CODE REVIEW

File: src/map/roe.cpp [hunk 1]
```diff
@@ -557,4 +601,60 @@ namespace roeutils
...
+            if (unityEval[i + 1].second < unityEval[i].second)
+            {
+                currentRank = currentRank + rankGap;
+                rankGap     = 1;
+            }
+            else
+            {
+                rankGap++;
+            }
+        }
```
Comment (cocosolos) [inline src/map/roe.cpp]: Would we be able to shorten this to something like:
```cpp
        const char* unityQuery = "SELECT leader, members_prev, points_prev FROM unity_system ORDER BY CASE WHEN members_prev = 0 THEN 0 ELSE points_prev/members_prev END DESC;";
        int32       ret        = Sql_Query(SqlHandle, unityQuery);

        if (ret != SQL_ERROR && Sql_NumRows(SqlHandle) != 0)
        {
            uint8 currentRank = 1;
            int32 prev_eval   = 0;
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                int32 members  = Sql_GetIntData(SqlHandle, 1);
                int32 points   = Sql_GetIntData(SqlHandle, 2);
                int32 new_eval = members == 0 ? 0 : points / members;
                if (new_eval < prev_eval)
                {
                    currentRank++;
                }
                prev_eval = new_eval;
                roeutils::RoeSystem.unityLeaderRank[Sql_GetIntData(SqlHandle, 0) - 1] = currentRank;
            }
        }
```
It looks like we're doing a lot of extra work when the SQL could sort it for us.

File: src/map/utils/charutils.cpp [hunk 2]
```diff
@@ -5331,6 +5333,37 @@ namespace charutils
...
+            // Do not allow subracting points with this function
+            return;
+        }
+
+        AddPoints(PChar, "unity_accolades", amount, max);
+        AddPoints(PChar, "current_accolades", amount, std::numeric_limits<int32>::max()); // Do not cap current_accolades
+        Sql_Query(SqlHandle, rankingQuery, evalPoints, PChar->profile.unity_leader);
+
+        PChar->pushPacket(new CCharStatsPacket(PChar));
+    }
```
Comment (cocosolos) [inline src/map/utils/charutils.cpp]: How do we feel about rolling this into the existing AddPoints with a condition similar to how we handle `spark_of_eminence`?

File: src/map/roe.cpp (hunk 1 above)
Comment (ghost) [reply src/map/roe.cpp]: Agreed.  I'm an SQL potato, and will update and test from the above suggestions.

Comment (cocosolos) [reply src/map/roe.cpp]: I believe the above will have a unity with 0 members ranked the same as unities with 0 points but some members. If you want members to count for something can change to this:
```cpp
        const char* unityQuery = "SELECT leader, members_prev, points_prev FROM unity_system ORDER BY CASE WHEN members_prev = 0 THEN -1 ELSE points_prev/members_prev END DESC;";
```
and
```cpp
                int32 new_eval = members == 0 ? -1 : points / members;
```

File: src/map/utils/charutils.cpp (hunk 2 above)
Comment (ghost) [reply src/map/utils/charutils.cpp]: So long as we don't have a case where we'd want to add accolades to a user without affecting ranking, I'm on board with that.  Any thoughts?

Comment (cocosolos) [reply src/map/utils/charutils.cpp]: You could bundle that in a check couldn't you, like:
```cpp
    void AddPoints(CCharEntity* PChar, const char* type, int32 amount, int32 max)
    {
        const char* Query = "UPDATE char_points SET %s = GREATEST(LEAST(%s+%d, %d), 0) WHERE charid = %u;";

        Sql_Query(SqlHandle, Query, type, type, amount, max, PChar->id);

        if (strcmp(type, "unity_accolades") == 0)
        {
            float       evalPoints    = static_cast<float>(amount) / 1000;
            const char* rankingQuery  = "UPDATE unity_system SET points_current = points_current+%f WHERE leader=%d;";
            AddPoints(PChar, "current_accolades", amount, std::numeric_limits<int32>::max()); // Do not cap current_accolades
            Sql_Query(SqlHandle, rankingQuery, evalPoints, PChar->profile.unity_leader);

            PChar->pushPacket(new CCharStatsPacket(PChar));
        }
        else if (strcmp(type, "spark_of_eminence") == 0)
        {
            PChar->pushPacket(new CRoeSparkUpdatePacket(PChar));
        }
    }
```

File: src/map/roe.cpp (hunk 1 above)
Comment (cocosolos) [reply src/map/roe.cpp]: Actually, I think you can even change the query to `SELECT leader, members_prev, points_prev, CASE WHEN members_prev = 0 THEN 0 ELSE points_prev/members_prev END AS eval FROM unity_system ORDER BY eval DESC;` and then use `int32 new_eval = Sql_GetIntData(SqlHandle, 3);`

Edit: Something like this:
```cpp
        const char* unityQuery = "SELECT leader, CASE WHEN members_prev = 0 THEN 0 ELSE points_prev/members_prev END AS eval FROM unity_system ORDER BY eval DESC;";
        int32       ret        = Sql_Query(SqlHandle, unityQuery);

        if (ret != SQL_ERROR && Sql_NumRows(SqlHandle) != 0)
        {
            uint8 currentRank = 1;
            int32 prev_eval   = 0;
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                int32 new_eval = Sql_GetIntData(SqlHandle, 1);
                if (new_eval < prev_eval)
                {
                    currentRank++;
                }
                prev_eval = new_eval;
                roeutils::RoeSystem.unityLeaderRank[Sql_GetIntData(SqlHandle, 0) - 1] = currentRank;
            }
        }
```

File: src/map/utils/charutils.cpp (hunk 2 above)
Comment (ghost) [reply src/map/utils/charutils.cpp]: I think I was thinking more of a case revolving around a gm command; outside of that, there shouldn't be any situation where adding accolades doesn't impact ranking, so I'll make this change.

File: src/map/roe.cpp (hunk 1 above)
Comment (ghost) [reply src/map/roe.cpp]: Just getting coffeed up, but I remember why the ranking code is a bit convoluted.  For example, if two unities were tied for first, and three were tied for second, the result would be as follows:

1, 1, 3, 3, 3, 6, 7, ...

Comment (cocosolos) [reply src/map/roe.cpp]: Aye, the code I posted checks for ties and only increments rank if the evaluation is actually lower. The only place it gets weird is if members + no points is a higher rank than no members + no points, I'm not sure how it's supposed to work, but can be adjusted either way with the `-1` change mentioned.

Comment (ghost) [reply src/map/roe.cpp]: I would assume those should evaluate as the same rank ultimately, since they should be stored with the same rank (0 == 0).  Implementing this change after I finish capturing some more event ids!

Comment (cocosolos) [reply src/map/roe.cpp]: Also I'm not sure how it's supposed to be handled but you might want to consider float for prev_eval and new_eval and using Sql_GetFloatData.

Comment (ghost) [reply src/map/roe.cpp]: As long as the returned values are floored, I think we're fairly safe, but unfortunately the above example doesn't skip ranks when ties exist.  From the provided test scenario, expected result was 1, 1, 3, 3, 3, and received result 1, 1, 2, 2, 2

Important to note as well: As of the current implementation, this function is called in only two instances:
1. On server start
2. On weekly update

Comment (cocosolos) [reply src/map/roe.cpp]: Ah I misunderstood how it was supposed to work. You could add the same "rankGap" system you are currently using into the condensed version (I think this is correct).
```cpp
            uint8 currentRank = 1;
            uint8 rankGap     = 1;
            int32 prev_eval   = 0;
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                int32 new_eval = Sql_GetIntData(SqlHandle, 1);
                if (new_eval < prev_eval)
                {
                    currentRank = currentRank + rankGap;
                    rankGap     = 1;
                }
                else
                {
                    rankGap++;
                }
                prev_eval = new_eval;
                roeutils::RoeSystem.unityLeaderRank[Sql_GetIntData(SqlHandle, 0) - 1] = currentRank;
            }
```

Comment (ghost) [reply src/map/roe.cpp]: Thanks for that.  Updated the code to reflect your suggestions, though I did change rankGap initialized value to fix an obo that was occurring.

Comment (cocosolos) [reply src/map/roe.cpp]: Ah nice catch, looks good!

File: src/map/utils/charutils.cpp (hunk 2 above)
Comment (cocosolos) [reply src/map/utils/charutils.cpp]: Back to this real quick, do we still need the `addAccolades` binding? I'm just seeing it in the GM command and 1 other spot. Is the difference just the cap? Could use `player:addCurrency("unity_accolades", rewards["accolades"], CAP_CURRENCY_ACCOLADES)`?

Comment (ghost) [reply src/map/utils/charutils.cpp]: Agreed.  Since accolades are only granted as a result of RoE objective especially, there's no need for this binding or gm command.  Removed.

**Review by zach2good — APPROVED** [review]
Excellent work 👍
